src/AzureExtension/DevBox/DevBoxInstance.cs

```diff
@@ -541,6 +609,109 @@ public IAsyncOperation<ComputeSystemOperationResult> ModifyPropertiesAsync(strin
+                if (exitcode == ExitCodeSuccess)
+                {
+                    return new ComputeSystemOperationResult();
+                }
+            }
+
+            var errorString = $"DoPinActionAsync with location {location} and action {pinAction} failed with exitcode: {exitcode}";
+            _log.Error($"Error getting thumbnail for {DisplayName}");
+            return new ComputeSystemOperationResult(new NotSupportedException(), Resources.GetResource(Constants.DevBoxUnableToPerformOperationKey), errorString);
+        }).AsAsyncOperation();
+    }
+
+    public IAsyncOperation<ComputeSystemOperationResult> PinToStartMenuAsync()
+    {
+        return DoPinActionAsync("startMenu", "pin");
+    }
+
+    public IAsyncOperation<ComputeSystemOperationResult> UnpinFromStartMenuAsync()
+    {
+        return DoPinActionAsync("startMenu", "unpin");
+    }
+
+    public IAsyncOperation<ComputeSystemOperationResult> PinToTaskbarAsync()
+    {
+        return DoPinActionAsync("taskbar", "pin");
+    }
+
+    public IAsyncOperation<ComputeSystemOperationResult> UnpinFromTaskbarAsync()
+    {
+        return DoPinActionAsync("taskbar", "unpin");
+    }
+
+    public IAsyncOperation<ComputeSystemPinnedResult> GetPinStatusAsync(string location)
+    {
+        return Task.Run(() =>
+        {
+            var exitcode = ExitCodeInvalid;
+            var psi = new ProcessStartInfo();
+            psi.UseShellExecute = true;
+            psi.FileName = string.Format(CultureInfo.InvariantCulture, ProtocolPinString, location, "status", WorkspaceId, DisplayName, Username, Environment);
+            Process? p = Process.Start(psi);
+            if (p != null)
+            {
+                p.WaitForExit();
+                exitcode = p.ExitCode;
+                if (exitcode == ExitCodePinned)
+                {
+                    return new ComputeSystemPinnedResult(true);
+                }
+                else if (exitcode == ExitCodeUnpinned)
+                {
+                    return new ComputeSystemPinnedResult(false);
+                }
+            }
+
+            var errorString = $"GetPinStatusAsync from location {location} failed with exitcode: {exitcode}";
+            _log.Error($"Error getting thumbnail for {DisplayName}");
+            return new ComputeSystemPinnedResult(new NotSupportedException(), Resources.GetResource(Constants.DevBoxUnableToPerformOperationKey), errorString);
+        }).AsAsyncOperation();
+    }
+
+    public IAsyncOperation<ComputeSystemPinnedResult> GetIsPinnedToStartMenuAsync()
+    {
+        return GetPinStatusAsync("startmenu");
+    }
+
+    public IAsyncOperation<ComputeSystemPinnedResult> GetIsPinnedToTaskbarAsync()
+    {
+        return GetPinStatusAsync("taskbar");
+    }
+
+    public async void LoadWindowsAppParameters()
+    {
+        if (RemoteConnectionData is null)
+        {
+            await GetRemoteLaunchURIsAsync(new Uri(DevBoxState.Uri));

```

Comment:
I believe `new Uri` will throw if `DevBoxState.Uri` is not a valid Uri.  Has the Uri been validated by this point?

Reply:
Good question.  @huzaifa-d  do you know? And same question for RemoteConnectionData?.CloudPcConnectionUrl

Reply:
We deserialize the DevBoxMachineState here: https://github.com/microsoft/DevHomeAzureExtension/blob/4f3bfede9368f3710a37f1e877a750f34fbb7f43/src/AzureExtension/DevBox/DevBoxProvider.cs#L80

The DevBoxMachineState has a default value of string.Empty here: https://github.com/microsoft/DevHomeAzureExtension/blob/4f3bfede9368f3710a37f1e877a750f34fbb7f43/src/AzureExtension/DevBox/DevBoxJsonToCsClasses/DevBoxMachineState.cs#L14


We retrieve the URI directly as Json from the Dev Center API after this call https://github.com/microsoft/DevHomeAzureExtension/blob/4f3bfede9368f3710a37f1e877a750f34fbb7f43/src/AzureExtension/DevBox/DevBoxProvider.cs#L62

If there is an exception it would be because the dev center apis returned an invalid uri which would be against their API: See sample response: https://learn.microsoft.com/en-us/rest/api/devcenter/developer/dev-boxes/list-dev-boxes-by-user?view=rest-devcenter-developer-2024-02-01&tabs=HTTP

In the end we have a try/catch over the method that this call is made from

Reply:
Great, thanks